Fix SF10a driver. Add support for SF11c and rename to SF1xx

svpcom opened the pull request:
Current i2c driver for SF10a (also for all SF1XX family) is non-functional and violates device datasheet.  I've fixed it and added support for SF11C altimeter

LorenzMeier commented:
Pretty cool, thanks!

jgoppert commented on src/drivers/sf1xx/sf1xx.cpp:
```diff
@@ -594,24 +593,17 @@ SF10A::cycle()
-	if (OK != measure()) {
-		DEVICE_DEBUG("measure error lidar");
-	}
-
```
It looks like you removed the measure function on accident.

svpcom replied:
No, it not required according to datasheet. You write '\0' byte once to start measurements and then only poll for updates.

```
Values in the SF11 are held at specific locations and once a value has been read or written the location remains active until changed.
This is useful to simplify the code needed for the continuous monitoring of a value. For example, to update the distance
continuously, set the distance location once then read the value in a loop like the Arduino code below:
```
![ds](https://cloud.githubusercontent.com/assets/503830/17240131/6522160a-557a-11e6-8003-c771c67d5069.png)

jgoppert replied:
Interesting, we might need to worry about the edge case when measure fails, we would want to try to set the bit again.

svpcom replied:
According to datasheet there the only case of stopping measurement update if someone write other that '\0' to device. In case of device reset (due to power failure or firmware bug) we can write '\0' in case of measurement timeout.

jgoppert replied:
OK, agree with your approach, looks good. Have you tested it yet? Would be good to see some logs.

svpcom replied:
Yes, i've tested it by hot-remove and hot-plug sensor and ensure that uORB announcements restarts back.

jgoppert commented on src/drivers/sf1xx/sf1xx.cpp:
```diff
@@ -75,18 +76,18 @@
-#define SF10A_MAX_DISTANCE 	(25.0f)
+#define SF1XX_MIN_DISTANCE 	(0.01f)
+#define SF1XX_MAX_DISTANCE 	(120.0f)
 
```
Is 120m valid for all sf1xx series? Surely there are differences, how best to handle these? Maybe if SENS_EN_SF10A is set it can tell the driver to start the sf1xx driver with the correct parameters in the init script.

svpcom replied:
In SF11/c datasheet specify that sensor returns 130m in case of signal loss. In SF10/x there are no such section.Also in SF10/x there are no (officially) i2c command to identify device. So if someone has sf10/x lidar it can test this case. I'll add SENS_EN_SF1XX to select proper hw runtime if needed

jgoppert commented on src/drivers/sf1xx/sf1xx.cpp:
```diff
@@ -75,18 +76,18 @@
 
-#define SF10A_CONVERSION_INTERVAL 		31250		// Maximum rate according to datasheet is 32Hz
+#define SF1XX_CONVERSION_INTERVAL 		50000		// Maximum rate according to datasheet is 20Hz
 
```
Similar comment to max distance above, how to handle changes for each model?

svpcom replied:
SF10/{a,b} has 32Hz rate and SF11/c - 20Hz. So we can leave 20Hz safely

svpcom commented:
@jgoppert I've added SENS_EN_SF1XX with following possible values:
 * 0 - disable
 * 1 - SF10a
 * 2 - SF10b
 * 3 - SF10c
 * 4 - SF11c

Also i've fixed long standing bug with missed ``orb_unadvertise`` on driver stop which leads to broken publish after driver stop/start cycle. Also there no ``close`` calls to opened fd's. 

I've found that many of the device drivers forgets to:
1. ``orb_unadvertise`` on stop (such driver will broken after restart).
2.  close file descriptors.

jgoppert commented on src/modules/sensors/sensor_params.c:
```diff
@@ -3009,6 +3009,22 @@ PARAM_DEFINE_INT32(SENS_EN_LL40LS, 0);
+ * @value 2 SF10/b (50m)
+ * @value 3 SF10/c (100m)
+ * @value 4 SF11/c (120m)
+ */
```
Nice!

jgoppert commented on src/modules/sensors/sensor_params.c:
```diff
@@ -3009,6 +3009,22 @@ PARAM_DEFINE_INT32(SENS_EN_LL40LS, 0);
+ * @max 4
+ * @group Sensor Enable
+ * @value 0 Disabled
+ * @value 1 SF10/a (25m)
```
probably can leave off the (25m) because it seems you are also setting conversion interval

jgoppert commented:
If you have time to do a quick flight with the driver change I would appreciate it. I have a vehicle flying with an sf10a and don't want to find a glitch in flight.

svpcom commented:
I'll try to check it in hands (without props) today.
Also please do ground test too to be sure that sf10a also stream measurements (its datasheet say only few words about i2c protocol)

svpcom commented:
on the ground:
http://logs.uaventure.com/view/oWJUdudR9eQzFy7nMCjhrd

in the hands:
http://logs.uaventure.com/view/ZYSs7xfuXFxZnqvN3uAYWe

ecmnet commented:
@svpcom I am curious whether the readings are correct. This is what your first log (on the ground) shows as raw readings:

![chart](https://cloud.githubusercontent.com/assets/7580193/17269125/b0266d56-5640-11e6-8cdd-d773c36b75ec.png)

svpcom commented:
http://logs.uaventure.com/view/ZWrto8kKKyszWzNswzQuyN
this is log from my today's flight

svpcom commented:
is the second plot from sf10a lidar or it is from sf11c? sf11c is a long-range and has more fluctuations in short range measures.

svpcom commented:
@ecmnet Can't you compare old and new sf1xx drivers on sf10a lidar? I'm unable to compare old because it is not work on my sf11c.

svpcom commented:
@ecmnet Also sf11c has builtin median filter (disabled by default). I've enabled it and it removed spikes. So now it report altitude with +- 1cm deviance

ecmnet commented:
@svpcom I will do a comparison as soon as I finde some time. Some experiments with SF10a two years ago showed, that I2C communication (e.g. the clock speed and EMV) could influence the device showing random or repeating spikes. Could you share a latest log?

svpcom commented:
@ecmnet 
http://logs.uaventure.com/view/8kBhiED5fLrqaDpoZLNRmN
sf11c with internal median filter and speed vs distance measurement mode.
max flight altitude was ~ 150m, terrain - high grass.

svpcom commented:
It seems i've found issue of spikes -- i2c address collision with rgbled driver

LorenzMeier commented:
@svpcom What configuration and wiring is required on the sensor side to make this work? Can you share a picture? Ideally you would send a PR for PX4/Devguide.

svpcom commented:
![pixhawk_rangefinder_sf1xx_i2c](https://cloud.githubusercontent.com/assets/503830/17291191/7c81bb40-57f3-11e6-8ea4-4eb536f3f333.jpg)

Before this you should connect to sensor via usb (it has internal usb2serial converter),  run terminal, press 'space' and check that i2c address == 0x66. Newer sensor versions already have 0x66 preconfigured.

ecmnet commented on src/drivers/sf1xx/sf1xx.cpp:
```diff
@@ -77,7 +77,7 @@
 /* Configuration Constants */
 #define SF1XX_BUS 		PX4_I2C_BUS_EXPANSION
-#define SF1XX_BASEADDR 	0x55
+#define SF1XX_BASEADDR 	0x66
```
This means that the default address of SF10a is not used anymore. Additionally, I wonder why 0x55 did no give any spikes earlier then.

svpcom replied:
You could change i2c address in SF10a the same way as in SF11c (via usb menu)

svpcom commented:
@LorenzMeier Is it possible to merge into upcoming 1.4.1 release?

LorenzMeier commented:
Generally yes, but I would love to have some cross-testing from other users in this thread.

svpcom commented:
@jgoppert , @ecmnet  Could you test new SF1xx driver with SF10a lidar?

ecmnet commented:
@svpcom I've planned some tests for today.

ecmnet commented:
@svpcom @LorenzMeier @jgoppert Tested the new driver with SF10a and it works fine for me. I also did a flight test with V1.4.1RC1: http://logs.uaventure.com/view/w6HFPnTK8MkeQA3HJYg6J5. It would be great to see it in V1.4.1.

svpcom commented:
@LorenzMeier Changes was cross-tested by @ecmnet, documentation already merged to devguide. I think it ready for v1.4.1 merge

LorenzMeier commented:
It's rebased on master, thanks!
